sys/linux/socket_netlink_route_sched.txt

```diff
@@ -95,6 +135,136 @@ q_cbs_options [
+] [varlen]
+
+q_dsmark_options [
+#TODO: hweight32(indices) should equal to 1
+	TCA_DSMARK_INDICES		nlattr[TCA_DSMARK_INDICES, int16[0:32]]
+	TCA_DSMARK_DEFAULT_INDEX	nlattr[TCA_DSMARK_DEFAULT_INDEX, int16]
+	TCA_DSMARK_SET_TC_INDEX		nlattr[TCA_DSMARK_SET_TC_INDEX, void]
+] [varlen]
+
+q_fq_options [
+	TCA_FQ_PLIMIT			nlattr[TCA_FQ_PLIMIT, int32]
+	TCA_FQ_FLOW_PLIMIT		nlattr[TCA_FQ_FLOW_PLIMIT, int32]
+	TCA_FQ_QUANTUM			nlattr[TCA_FQ_QUANTUM, int32]
+	TCA_FQ_INITIAL_QUANTUM		nlattr[TCA_FQ_INITIAL_QUANTUM, int32]
+	TCA_FQ_RATE_ENABLE		nlattr[TCA_FQ_RATE_ENABLE, int32[0:1]]
+	TCA_FQ_FLOW_DEFAULT_RATE	nlattr[TCA_FQ_FLOW_DEFAULT_RATE, int32]
+	TCA_FQ_FLOW_MAX_RATE		nlattr[TCA_FQ_FLOW_MAX_RATE, int32]
+	TCA_FQ_BUCKETS_LOG		nlattr[TCA_FQ_BUCKETS_LOG, int32[0:32]]
+	TCA_FQ_FLOW_REFILL_DELAY	nlattr[TCA_FQ_FLOW_REFILL_DELAY, int32]
+	TCA_FQ_ORPHAN_MASK		nlattr[TCA_FQ_ORPHAN_MASK, int32]
+	TCA_FQ_LOW_RATE_THRESHOLD	nlattr[TCA_FQ_LOW_RATE_THRESHOLD, int32]
+] [varlen]
+
+q_fq_codel_options [
+	TCA_FQ_CODEL_TARGET		nlattr[TCA_FQ_CODEL_TARGET, int32]
+	TCA_FQ_CODEL_LIMIT		nlattr[TCA_FQ_CODEL_LIMIT, int32]
+	TCA_FQ_CODEL_INTERVAL		nlattr[TCA_FQ_CODEL_INTERVAL, int32]
+	TCA_FQ_CODEL_ECN		nlattr[TCA_FQ_CODEL_ECN, int32[0:1]]
+	TCA_FQ_CODEL_FLOWS		nlattr[TCA_FQ_CODEL_FLOWS, int32]
+	TCA_FQ_CODEL_QUANTUM		nlattr[TCA_FQ_CODEL_QUANTUM, int32]
+	TCA_FQ_CODEL_CE_THRESHOLD	nlattr[TCA_FQ_CODEL_CE_THRESHOLD, int32]
+	TCA_FQ_CODEL_DROP_BATCH_SIZE	nlattr[TCA_FQ_CODEL_DROP_BATCH_SIZE, int32]
+	TCA_FQ_CODEL_MEMORY_LIMIT	nlattr[TCA_FQ_CODEL_MEMORY_LIMIT, int32]
+] [varlen]
+
+# TODO: we should not have TCA_GRED_PARMS and TCA_GRED_STAB when do init
+q_gred_options [
+	TCA_GRED_PARMS	nlattr[TCA_GRED_PARMS, tc_gred_qopt]
+	TCA_GRED_STAB	nlattr[TCA_GRED_STAB, array[int8, 256]]
+	TCA_GRED_DPS	nlattr[TCA_GRED_DPS, tc_gred_sopt]
+	TCA_GRED_MAX_P	nlattr[TCA_GRED_MAX_P, int32]
+	TCA_GRED_LIMIT	nlattr[TCA_GRED_LIMIT, int32]
+] [varlen]
+
+q_hhf_options [
+	TCA_HHF_BACKLOG_LIMIT	nlattr[TCA_HHF_BACKLOG_LIMIT, int32]
+	TCA_HHF_QUANTUM		nlattr[TCA_HHF_QUANTUM, int32]
+	TCA_HHF_HH_FLOWS_LIMIT	nlattr[TCA_HHF_HH_FLOWS_LIMIT, int32]
+	TCA_HHF_RESET_TIMEOUT	nlattr[TCA_HHF_RESET_TIMEOUT, int32]
+	TCA_HHF_ADMIT_BYTES	nlattr[TCA_HHF_ADMIT_BYTES, int32]
+	TCA_HHF_EVICT_TIMEOUT	nlattr[TCA_HHF_EVICT_TIMEOUT, int32]
+	TCA_HHF_NON_HH_WEIGHT	nlattr[TCA_HHF_NON_HH_WEIGHT, int32]
+] [varlen]
+
+q_htb_options [
+	TCA_HTB_INIT		nlattr[TCA_HTB_INIT, tc_htb_glob]
+	TCA_HTB_DIRECT_QLEN	nlattr[TCA_HTB_DIRECT_QLEN, int32]
+] [varlen]
+
+q_mqprio_options [
+	TCA_MQPRIO_MODE		nlattr[TCA_MQPRIO_MODE, flags[tc_mqprio_modes, int32]]
+	TCA_MQPRIO_SHAPER	nlattr[TCA_MQPRIO_SHAPER, flags[tc_mqprio_shapers, int32]]
+	TCA_MQPRIO_MIN_RATE64	array[nlattr[TCA_MQPRIO_MIN_RATE64, int64], 16]
+	TCA_MQPRIO_MAX_RATE64	array[nlattr[TCA_MQPRIO_MAX_RATE64, int64], 16]
+] [varlen]
+
+q_netem_options [
+	TCA_NETEM_CORR		nlattr[TCA_NETEM_CORR, tc_netem_corr]
+# TODO: should be distribution files
```

Comment:
What does this mean? Does this ask for more descriptions? Or improvements in descriptions syntax and capabilities? It would be useful if all readers could understand it.

Reply:
TCA_NETEM_DELAY_DIST data should be a netem distribution file. I will update the TODO description.

Reply:
Oh, I also just figure out, it should be the data in /usr/lib64/tc/*.dist. So I think we could set it like
`TCA_NETEM_DELAY_DIST    nlattr[TCA_NETEM_DELAY_DIST, array[int16, 4096]]` or a flexible array like `array[int16, 0:16384]`,  16384 is from `MAX_DIST        (16*1024)` in iproute2.

Reply:
Do we really need it that large? I am somewhat concerned with size of programs and reproducers.
Also let's use int8, array[int8] is treated in a special way with more efficient handling and serialization (like blob).

Reply:
What array size do you prefer. 4096 is the data size in dist files.

Reply:
Yes, that one. Can it be smaller?

Reply:
No idea... I haven't investigated what this file used for.

Reply:
The size can be anywhere from 0 to 16384 (NETEM_DIST_MAX) int16's:
https://elixir.bootlin.com/linux/v4.17-rc5/source/net/sched/sch_netem.c#L733

From the code it does not look too important to pass huge sizes. Let's start with:
array[int8, 0:100]]